### PR TITLE
Add Get QueryGroup API Logic

### PR DESCRIPTION
**Pull request by ruai0511**
### Description
This change introduces the Get QueryGroup API which we will use to enforce node level resiliency as part of this RFC: https://github.com/opensearch-project/OpenSearch/issues/12342. The draft PR opened for the API specs is: https://github.com/opensearch-project/opensearch-api-specification/pull/356

The Get QueryGroup API schema is:
```
curl -XGET "localhost:9200/_wlm/query_group/"  // get all existing QueryGroups
curl -XGET "localhost:9200/_wlm/query_group/{name}"  // get the QueryGroup with name={name}
```

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [ ] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/rest/RestGetQueryGroupAction.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+     * The list of {@link Route}s that this RestHandler is responsible for handling.
+     */
+    @Override
+    public List<Route> routes() {
+        return List.of(new Route(GET, "_query_group/{name}"), new Route(GET, "_query_group/"));
+    }
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        String name = request.param("name");
```
Comment (kaushalmahi12) [inline plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/rest/RestGetQueryGroupAction.java:56]: I am assuming this API should demonstrate the following behavior
* If name not provided return all the query groups.
* If name is provided then return the matching query group.

Does the API return complete match query groups or prefix match or substring match ?

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/service/QueryGroupPersistenceService.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    List<QueryGroup> getFromClusterStateMetadata(String name, ClusterState currentState) {
+        Map<String, QueryGroup> currentGroups = currentState.getMetadata().queryGroups();
+        if (name == null || name.isEmpty()) {
+            return new ArrayList<>(currentGroups.values());
+        }
+        List<QueryGroup> resultGroups = new ArrayList<>();
+        for (QueryGroup group : currentGroups.values()) {
+            if (group.getName().equals(name)) {
+                resultGroups.add(group);
+                break;
```
Comment (kaushalmahi12) [inline plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/service/QueryGroupPersistenceService.java:107]: Should we return all the substring matched query groups ?

File: plugins/workload-management/src/test/java/org/opensearch/plugin/wlm/action/GetQueryGroupResponseTests.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import org.opensearch.common.io.stream.BytesStreamOutput;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.rest.RestStatus;
+import org.opensearch.test.OpenSearchTestCase;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class GetQueryGroupResponseTests extends OpenSearchTestCase {
```
Comment (kaushalmahi12) [inline plugins/workload-management/src/test/java/org/opensearch/plugin/wlm/action/GetQueryGroupResponseTests.java:21]: can we add a toXContent test case to demonstrate how should the response will look like in JSON format

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/rest/RestGetQueryGroupAction.java (hunk 1 above)
Comment (ruai0511) [reply plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/rest/RestGetQueryGroupAction.java:56]: I see [get index api](https://opensearch.org/docs/latest/api-reference/index-apis/get-index/) and [get settings api](https://opensearch.org/docs/latest/api-reference/index-apis/get-settings/) both return the exact match

**Review by kaushalmahi12 — APPROVED** [review]

**Comment by kaushalmahi12** [comment]
Can we increase the line coverage ? rest of it looks fine

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/GetQueryGroupRequest.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+     * @param in - A {@link StreamInput} object
+     */
+    public GetQueryGroupRequest(StreamInput in) throws IOException {
+        super(in);
+        name = in.readOptionalString();
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        return null;
```
Comment (jainankitk) [inline plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/GetQueryGroupRequest.java:45]: Do we want to validate the name length, allowed characters before getting the query group from cluster manager?

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/GetQueryGroupResponse.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Response for the get API for QueryGroup
+ *
+ * @opensearch.experimental
+ */
+public class GetQueryGroupResponse extends ActionResponse implements ToXContent, ToXContentObject {
+    private final List<QueryGroup> queryGroups;
```
Comment (jainankitk) [inline plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/GetQueryGroupResponse.java:29]: In what scenarios, Get API might return more than one query groups? While I doubt that is possible, but if it were, it should be BatchGet API?

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/TransportGetQueryGroupAction.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        super(GetQueryGroupAction.NAME, transportService, actionFilters, GetQueryGroupRequest::new);
+        this.threadPool = threadPool;
+        this.queryGroupPersistenceService = queryGroupPersistenceService;
+    }
+
+    @Override
+    protected void doExecute(Task task, GetQueryGroupRequest request, ActionListener<GetQueryGroupResponse> listener) {
+        threadPool.executor(ThreadPool.Names.GENERIC)
+            .execute(() -> queryGroupPersistenceService.getFromClusterStateMetadata(request.getName(), listener));
+    }
```
Comment (jainankitk) [inline plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/TransportGetQueryGroupAction.java:56]: We should be using ClusterManager action for getting the information from latest ClusterState?

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        }
+    }
+
+    /**
+     * Get the QueryGroup with the specified name and generate response
+     * @param name - the QueryGroup name we are getting
+     * @param listener - ActionListener for GetQueryGroupResponse
+     */
+    public void getFromClusterStateMetadata(String name, ActionListener<GetQueryGroupResponse> listener) {
+        List<QueryGroup> resultGroups = getQueryGroupsFromClusterState(name, clusterService.state());
```
Comment (jainankitk) [inline plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java:103]: This is the cluster service on request receiving the request which might be lagging the cluster state version. We should fetch this information on master node to ensure we get the query group from latest information. This also breaks the eventually consistency constraint, since the data nodes might be temporarily inconsistent which can reflect in the query group information that we share with the client. First invocation might show a query group, the next might not show query group

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/GetQueryGroupResponse.java (hunk 5 above)
Comment (ruai0511) [reply plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/GetQueryGroupResponse.java:29]: `_wlm/query_group/{name} `returns the single querygroup with {name}, and` _wlm/query_group/ `returns all the querygroups in the system

Comment (jainankitk) [reply plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/GetQueryGroupResponse.java:29]: We are probably overloading `get` and `list` together in single API. I am wondering if there is any precedent for that in Opensearch.

Comment (ruai0511) [reply plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/GetQueryGroupResponse.java:29]: I came across this: https://opensearch.org/docs/latest/api-reference/index-apis/get-settings/, where `GET /<target-index>/_settings` retrieves all the settings for the specified index and `GET /<target-index>/_settings/<setting>` retrieves a specific setting for the specified index

**Comment by jainankitk** [comment]
@ruai0511 - Can you rebase this from `main` given https://github.com/opensearch-project/OpenSearch/pull/14680 has already gone in?

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/GetQueryGroupRequest.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+     * @param in - A {@link StreamInput} object
+     */
+    public GetQueryGroupRequest(StreamInput in) throws IOException {
+        super(in);
+        name = in.readOptionalString();
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        if (name != null) {
```
Comment (jainankitk) [inline plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/GetQueryGroupRequest.java:46]: Do you want to throw error if name isEmpty? How is the user not specifying name interpreted? `null` or `empty`?

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/TransportGetQueryGroupAction.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    @Override
+    protected ClusterBlockException checkBlock(GetQueryGroupRequest request, ClusterState state) {
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
+    }
+
+    @Override
+    protected void clusterManagerOperation(GetQueryGroupRequest request, ClusterState state, ActionListener<GetQueryGroupResponse> listener)
+        throws Exception {
+        String name = request.getName();
+        List<QueryGroup> resultGroups = QueryGroupPersistenceService.getFromClusterStateMetadata(name, state);
```
Comment (jainankitk) [inline plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/TransportGetQueryGroupAction.java:89]: nit: use final keyword

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/TransportGetQueryGroupAction.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+        if (resultGroups.isEmpty() && name != null && !name.isEmpty()) {
+            logger.warn("No QueryGroup exists with the provided name: {}", name);
+            throw new IllegalArgumentException("No QueryGroup exists with the provided name: " + name);
```
Comment (jainankitk) [inline plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/TransportGetQueryGroupAction.java:93]: What is the http status code for `IllegalArgumentException`? We should probably return `ResourceNotFoundException`?

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/rest/RestGetQueryGroupAction.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+     * The list of {@link Route}s that this RestHandler is responsible for handling.
+     */
+    @Override
+    public List<Route> routes() {
+        return List.of(new Route(GET, "_wlm/query_group/{name}"), new Route(GET, "_wlm/query_group/"));
+    }
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        GetQueryGroupRequest getQueryGroupRequest = new GetQueryGroupRequest(request.param("name"));
```
Comment (jainankitk) [inline plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/rest/RestGetQueryGroupAction.java:56]: nit: final

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java [hunk 12]
```diff
@@ -192,6 +194,21 @@ private Map<ResourceType, Double> calculateTotalUsage(Map<String, QueryGroup> ex
...
 
+    /**
+     * Get the QueryGroups with the specified name from cluster state
+     * @param name - the QueryGroup name we are getting
+     * @param currentState - current cluster state
+     */
+    public static List<QueryGroup> getFromClusterStateMetadata(String name, ClusterState currentState) {
+        Map<String, QueryGroup> currentGroups = currentState.getMetadata().queryGroups();
+        if (name == null || name.isEmpty()) {
+            return new ArrayList<>(currentGroups.values());
```
Comment (jainankitk) [inline plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java:205]: Can we use more generic `Collection` interface instead of `List` to avoid this unnecessary conversion?

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java [hunk 13]
```diff
@@ -192,6 +194,21 @@ private Map<ResourceType, Double> calculateTotalUsage(Map<String, QueryGroup> ex
...
+        }
+        List<QueryGroup> resultGroups = new ArrayList<>();
+        currentGroups.values().stream().filter(group -> group.getName().equals(name)).findFirst().ifPresent(resultGroups::add);
+        return resultGroups;
```
Comment (jainankitk) [inline plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java:209]: return `currentGroups.values().stream().filter(group -> group.getName().equals(name)).findAny().stream().toArray();`. Having `findAny()` instead of `findFirst()` allows parallelization, although might not matter as much since our list should be fairly small due to limitations on the number of QueryGroups.

File: server/src/main/java/org/opensearch/cluster/metadata/QueryGroup.java [hunk 14]
```diff
@@ -111,6 +108,12 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeLong(updatedAtInMillis);
     }
 
+    public static void validateName(String name) {
+        if (name.length() > MAX_CHARS_ALLOWED_IN_NAME || name.isEmpty()) {
+            throw new IllegalArgumentException("QueryGroup.name shouldn't be empty or more than 50 chars long");
+        }
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/cluster/metadata/QueryGroup.java:114]: We should not assume `name` is non-null, especially when this is public method

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/GetQueryGroupRequest.java (hunk 8 above)
Comment (ruai0511) [reply plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/GetQueryGroupRequest.java:46]: if name is empty. the function QueryGroup.validateName will throw error

File: server/src/main/java/org/opensearch/cluster/metadata/Metadata.java [hunk 15]
```diff
@@ -1830,9 +1830,7 @@ static void validateDataStreams(SortedMap<String, IndexAbstraction> indicesLooku
             if (dsMetadata != null) {
                 for (DataStream ds : dsMetadata.dataStreams().values()) {
                     String prefix = DataStream.BACKING_INDEX_PREFIX + ds.getName() + "-";
-                    Set<String> conflicts = indicesLookup.subMap(prefix, DataStream.BACKING_INDEX_PREFIX + ds.getName() + ".") // '.' is the
-                                                                                                                               // char after
-                                                                                                                               // '-'
+                    Set<String> conflicts = indicesLookup.subMap(prefix, DataStream.BACKING_INDEX_PREFIX + ds.getName() + ".")
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/cluster/metadata/Metadata.java:1833]: Can we revert the changes in this file?

**Review by jainankitk — APPROVED** [review]
LGTM!

**Comment by jainankitk** [comment]
@mch2 - Can you help merge these changes?

**Comment by jed326** [comment]
> ❌ Gradle check result for [478d493](https://github.com/opensearch-project/OpenSearch/commit/478d4930e49d9910367083186fd000d0aef7ddd1): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/44598/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

It looks like the failures here are related to the PR

**Comment by jainankitk** [comment]
> > ❌ Gradle check result for [478d493](https://github.com/opensearch-project/OpenSearch/commit/478d4930e49d9910367083186fd000d0aef7ddd1): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/44598/)
> > Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?
> 
> It looks like the failures here are related to the PR

@ruai0511 - It seems the new test is failing. Can you address that?

```
> Task :plugins:workload-management:yamlRestTest FAILED

=== Standard output of node `node{:plugins:workload-management:yamlRestTest-0}` ===
```

**Comment by jed326** [comment]
Please rebase after https://github.com/opensearch-project/OpenSearch/pull/15270 was merged

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/TransportGetQueryGroupAction.java [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    @Override
+    protected void clusterManagerOperation(GetQueryGroupRequest request, ClusterState state, ActionListener<GetQueryGroupResponse> listener)
+        throws Exception {
+        final String name = request.getName();
+        final Collection<QueryGroup> resultGroups = QueryGroupPersistenceService.getFromClusterStateMetadata(name, state);
+
+        if (resultGroups.isEmpty() && name != null && !name.isEmpty()) {
+            logger.warn("No QueryGroup exists with the provided name: {}", name);
```
Comment (jed326) [inline plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/action/TransportGetQueryGroupAction.java:93]: nit: this seems redundant

File: plugins/workload-management/src/test/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceServiceTests.java [hunk 17]
```diff
@@ -6,6 +6,14 @@
...
  */
 
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
```
Comment (jed326) [inline plugins/workload-management/src/test/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceServiceTests.java:16]: nit: License looks duplicated

**Review by jed326 — APPROVED** [review]

**Comment by jed326** [comment]
> ❌ Gradle check result for [23bd5b5](https://github.com/opensearch-project/OpenSearch/commit/23bd5b59027bb50269ec957a8526813e0160e0e7): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/44641/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

@ruai0511 this looks like it's still related to the code in your PR

**Comment by jed326** [comment]
You should be able to validate the YAML tests locally following https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md#running-rest-tests-against-an-external-cluster

**Review by jed326 — APPROVED** [review]

**Comment by jed326** [comment]
@ruai0511 please open a manual backport. Thanks!

**Comment by jainankitk** [comment]
@jed326 - Probably need to merge the create backport first? - https://github.com/opensearch-project/OpenSearch/pull/15227

**Comment by jed326** [comment]
Makes sense, merged that one let me retrigger here

**Comment by jainankitk** [comment]
Nice, the trigger-bot succeeded after the create backport merge!
